### ISSUE TITLE
Functions are not supported in TO_DATE and TO_TIMESTAMP arguments

### ISSUE THREAD

**Issue by borcsokj** [jOOQ/jOOQ#12314]
### Expected behavior

```sql
SELECT TO_DATE((2020 + 1)::INTEGER || '-' || 2::INTEGER || '-' || 3::INTEGER, 'YYYY-MM-DD')
```

should work similar to TO_DATE containing string concatenation without functions, ie. 

```sql
SELECT TO_DATE((2021)::INTEGER || '-' || 2::INTEGER || '-' || 3::INTEGER, 'YYYY-MM-DD')
```

### Actual behavior

`org.jooq.impl.ParserException` thrown with stack trace:

```
org.jooq.impl.ParserException: Token ')' expected: [1:22] SELECT TO_DATE((2020 [*]+ 1)::INTEGER || '-' || 2::INTEGER || '-' || 3::INTEGER, 'YYYY-MM-DD')
    at org.jooq.impl.DefaultParseContext.expected(ParserImpl.java:13536
    at org.jooq.impl.DefaultParseContext.parse(ParserImpl.java:12826
    at org.jooq.impl.DefaultParseContext.parse(ParserImpl.java:12821
    at org.jooq.impl.DefaultParseContext.parseTerm(ParserImpl.java:7991
    at org.jooq.impl.DefaultParseContext.parseUnaryOps(ParserImpl.java:7142
    at org.jooq.impl.DefaultParseContext.parseExp(ParserImpl.java:7119
    at org.jooq.impl.DefaultParseContext.parseFactor(ParserImpl.java:7096
    at org.jooq.impl.DefaultParseContext.parseSum(ParserImpl.java:7042
    at org.jooq.impl.DefaultParseContext.parseNumericOp(ParserImpl.java:7027
    at org.jooq.impl.DefaultParseContext.parseCollated(ParserImpl.java:6968
    at org.jooq.impl.DefaultParseContext.parseConcat(ParserImpl.java:6958
    at org.jooq.impl.DefaultParseContext.parseField(ParserImpl.java:6871
    at org.jooq.impl.DefaultParseContext.parseFieldToDateIf(ParserImpl.java:10099
    at org.jooq.impl.DefaultParseContext.parseTerm(ParserImpl.java:7818
    at org.jooq.impl.DefaultParseContext.parseUnaryOps(ParserImpl.java:7142
    at org.jooq.impl.DefaultParseContext.parseExp(ParserImpl.java:7119
    at org.jooq.impl.DefaultParseContext.parseFactor(ParserImpl.java:7096
    at org.jooq.impl.DefaultParseContext.parseSum(ParserImpl.java:7042
    at org.jooq.impl.DefaultParseContext.parseNumericOp(ParserImpl.java:7027
    at org.jooq.impl.DefaultParseContext.parseCollated(ParserImpl.java:6968
    at org.jooq.impl.DefaultParseContext.parseConcat(ParserImpl.java:6958
    at org.jooq.impl.DefaultParseContext.parsePredicate(ParserImpl.java:5895
    at org.jooq.impl.DefaultParseContext.parseNot(ParserImpl.java:5797
    at org.jooq.impl.DefaultParseContext.parseAnd(ParserImpl.java:5787
    at org.jooq.impl.DefaultParseContext.parseOr(ParserImpl.java:5778
    at org.jooq.impl.DefaultParseContext.parseFieldOrRow(ParserImpl.java:6862
    at org.jooq.impl.DefaultParseContext.parseFieldOrRow(ParserImpl.java:6809
    at org.jooq.impl.DefaultParseContext.parseSelectField(ParserImpl.java:6805
    at org.jooq.impl.DefaultParseContext.parseSelectList(ParserImpl.java:6750
    at org.jooq.impl.DefaultParseContext.parseQueryPrimary(ParserImpl.java:1309
    at org.jooq.impl.DefaultParseContext.parseQueryTerm(ParserImpl.java:1213
    at org.jooq.impl.DefaultParseContext.parseQueryExpressionBody(ParserImpl.java:1179
    at org.jooq.impl.DefaultParseContext.parseSelect(ParserImpl.java:948
    at org.jooq.impl.DefaultParseContext.parseSelect(ParserImpl.java:943
    at org.jooq.impl.DefaultParseContext.parseQuery(ParserImpl.java:788
    at org.jooq.impl.DefaultParseContext.parseQuery0(ParserImpl.java:541
    at org.jooq.impl.ParserImpl.parseQuery(ParserImpl.java:376
    at org.jooq.impl.ParsingConnection$CacheValue.<init>(ParsingConnection.java:91
    at org.jooq.impl.ParsingConnection.lambda$translate$0(ParsingConnection.java:132
    at org.jooq.impl.Cache.run(Cache.java:103
    at org.jooq.impl.ParsingConnection.translate(ParsingConnection.java:128
    at org.jooq.impl.ParsingConnection.lambda$prepareAndBind$3(ParsingConnection.java:171
    at org.jooq.impl.ParsingStatement.prepareAndBind(ParsingStatement.java:330
    at org.jooq.impl.ParsingStatement.executeQuery(ParsingStatement.java:340
```

Similar parser exception thrown if other function used as parameter of TO_DATE, i.e. `Unknown function: [1:380] (TO_DATE((EXTRACT([*]YEAR from t02.C_DATE::DATE)::INTEGER)::INTEGER || '-02-03', 'YYYY-MM-DD'))`

### Steps to reproduce the problem

Run queries above.

### Versions

- jOOQ: 1.13.x+
- Java: Java 11
- Database (include vendor): PostgreSQL 9+, Hsqldb
- OS: Ubuntu 20.04.2 LTS
- JDBC Driver (include name if inofficial driver): org.postgresql:postgresql:42.2.12, org.hsqldb:hsqldb:2.3.4

**Comment by lukaseder** [comment]
Thank you very much for your report. I can reproduce this, will look into this right away.

**Comment by lukaseder** [comment]
A few other functions are affected as well. The problem is that we have a wrong optimisation in the parser that expects expressions to be of a certain type, e.g. the first argument of `TO_DATE()` must be of some string type. But the current implementation prevents all non-string typed, nested expressions in that argument, even if the final type is string again, due to the concatenation.

I'll check if we can remove that logic for all functions.

**Comment by lukaseder** [comment]
This is the same issue as https://github.com/jOOQ/jOOQ/issues/11703. Closing as a duplicate. Will increase the priority of #11703

**Comment by lukaseder** [comment]
https://github.com/jOOQ/jOOQ/issues/11703 has been fixed for jOOQ 3.16.0 and 3.15.2, which should be released soon. Thanks again for your report.

**Comment by borcsokj** [comment]
Thank you for your quick fix.
